spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/autoconfig/CloudSqlJdbcUrlResolver.java

```diff
@@ -0,0 +1,80 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.sql.autoconfig;
+
+import java.io.IOException;
+
+import com.google.api.services.sqladmin.SQLAdmin;
+
+import org.springframework.cloud.gcp.sql.GcpCloudSqlProperties;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Utility class that resolves the JDBC URL from a number of parameters.
+ */
+class CloudSqlJdbcUrlResolver {
+
+	private final String projectId;
+
+	private final String templateUrl;
+
+	private final GcpCloudSqlProperties properties;
+
+	private final SQLAdmin sqlAdmin;
+
+	CloudSqlJdbcUrlResolver(String templateUrl, String projectId,
+			GcpCloudSqlProperties properties, SQLAdmin sqlAdmin) {
+		this.templateUrl = templateUrl;
+		this.projectId = projectId;
+		this.properties = properties;
+		this.sqlAdmin = sqlAdmin;
+	}
+
+	String resolveJdbcUrl() {
+		Assert.hasText(this.projectId, "A project ID must be provided.");
+		Assert.hasText(this.properties.getDatabaseName(), "A database name must be provided.");
+
+		String instanceConnectionName = this.properties.getInstanceConnectionName();
+		if (StringUtils.isEmpty(instanceConnectionName)) {
+			Assert.hasText(this.properties.getInstanceName(),
+					"An instance name is required, or specify an instance connection name explicitly");
+			String region = this.properties.getRegion();
+			if (StringUtils.isEmpty(region)) {
+				try {
+					region = determineRegion(this.projectId, this.properties.getInstanceName());
```

Comment:
I see the main reason of this class to determine a `region` against provided `projectId` and `instanceName`.
Why can't we be more strict and really require the `region` from end-user all the time? If `instanceConnectionName` isn't provided, of course.

Reply:
It's a feature to get people started even more easily. People tend to memorize the instance name better than its region (at least, I do). It's the difference between having to go to the Cloud Console to get an instance's region or skipping that step completely.

Although we should point out in the docs that this is not recommended for production environments.

Reply:
Yeah... I would really prefer to ask end-user to provide it rather than have such a "creepy" hooks during application start up... But let it be if you're OK to support this code in the future. Only the problem I'd like to see is `static` bean for the `SqlAdmin` do not load any other beans for nothing during this initialization phase.

Reply:
Apologies, I didn't understand what you want with the `SQLAdmin` bean.

Reply:
OK. Never mind. It doesn't help because that `DefaultCloudSqlJdbcInfoProvider` fully depends on the `SQLAdmin` bean. Although again: the last one is used only to determine a string for the `region`. Where it would be much easier just to request it from the user and don't care about `SQLAdmin` overhead at all.

Reply:
Judging from some talks, removing region auto-discovery is where we're headed.

Reply:
OK. And let me guess: this issue is exactly about that: https://github.com/spring-cloud/spring-cloud-gcp/issues/142.

Reply:
You got it :)
I think I addressed all your comments, but let me know if I missed anything.